addons/crm_addons/controllers/main.py

```diff
@@ -0,0 +1,387 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+import logging
+
+from addons.iap import InsufficientCreditError
+from odoo import http, _,  tools
+
+from odoo.http import request
+import json
+import base64
+import odoo
+import werkzeug
+import hmac
+from hashlib import sha256
+import time
+
+_logger = logging.getLogger(__name__)
+
+
+class CrmAddonsController(http.Controller):
+
+    @staticmethod
+    def _authorize(request):
+        token = dict(request.httprequest.headers).get('Authorization')
+        b64_encoded_header, b64_encoded_payload, b64_encoded_signature = token.split('.')
+
+        # Check signature.
+        database_secret = request.env['ir.config_parameter'].sudo().get_param('database.secret')
+        signature = hmac.new(database_secret.encode(),
+                             ('%s.%s' % (b64_encoded_header, b64_encoded_payload)).encode(),
+                             sha256).hexdigest()
+        signature_provided = base64.b64decode(
+            b64_encoded_signature).decode()  # decode b64, then decode the resulting binary string
+        if not hmac.compare_digest(signature, signature_provided):
+            return False
+
+        payload_string = base64.b64decode(b64_encoded_payload).decode()
+        payload_json = json.loads(payload_string)
+
+        user_id = payload_json.get('user_id')
+        expiration_time = payload_json.get('time_valid')
+
+        if not user_id or not expiration_time:
+            return False
+
+        if int(expiration_time) < time.time():
+            return False
+
+        user = request.env['res.users'].browse(user_id)
+        if not user:
+            return False
+
+        request.env.user = user.id
+        return True
+
+    @http.route('/crm/addons/login', type='http', auth="none")
+    def web_login(self, redirect=None, **kw):
```

Comment:
This is disturbing. Why does this need its own ad-hoc non-standard login system when even oauth2 on odoo.com apparently doesn't?